source/common/stats/thread_local_store.cc

```diff
@@ -75,6 +91,36 @@ void ThreadLocalStoreImpl::shutdownThreading() {
   shutting_down_ = true;
 }
 
+void ThreadLocalStoreImpl::mergeHistograms() {
+  if (!shutting_down_) {
+    std::unique_lock<std::mutex> lock(lock_);
+    // TODO: Wait till all threads have been checked-in
+    tls_->runOnAllThreads([this]() -> void {
+      for (ScopeImpl* scope : scopes_) {
+        for (auto histogram : tls_->getTyped<TlsCache>().scope_cache_[scope].histograms_) {
+          TlsHistogramSharedPtr tls_histogram_ptr =
+              std::dynamic_pointer_cast<ThreadLocalHistogramImpl>(histogram.second);
+          tls_histogram_ptr->beginMerge();
+        }
+      }
+    });
+    if (main_thread_dispatcher_) {
+      main_thread_dispatcher_->post([this]() -> void { mergeInternal(); });
+    }
+  }
+}
+
+void ThreadLocalStoreImpl::mergeInternal() {
+  std::unique_lock<std::mutex> lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
```

Comment:
I would use `histograms()` here, which will acquire the lock and quickly sort through histograms that need merging. Then, you can operate on each individual histogram safely without any lock since the merging process will take a larger amount of time.